docs/zkapps/advanced-snarkyjs.mdx

```diff
@@ -291,3 +291,170 @@ We imagine that one of the most important use cases for actions is to enable con
+
+Lets look at our data structure first. Imagine wanting to map a player's id to score points, so this:
+
+```
+0: 5 points
+1: 3 points
+2: 0 points
+3: 8 points
+... : ...
+7: 2 points
+```
+
+#### Implementing the smart contract
+
+Now it's time to look at what our leaderboard zkApp might look like. We want to have on-chain state that points to the off-chain Merkle tree - we will call this variable the `root`.
+
+<Alert kind="note">
+
+Sometimes the variable `root` is also called commitment, because we commit to something!
+
+</Alert>
+
+Additionally, we want to store a variable `z` which is the hash of the value a player has to guess: `H(guess) = z`
+
+The first method allows a player to make a guess, and if the guess is correct, the player will gain one point.
+The method takes the players guess and hashes it, it then checks if the hash H(guess) equals the on-chain state z and if that's the case,
+then the player gains one point on the score board.
+
+We will also need a second method that takes care of the reward. It checks if the players score is over a threshold and pays out a reward if that's the case.
+This method will also have to verify our Merkle witness and check it if it matches the on-chain stored Merkle root!
+
+<Alert kind="note">
+
+A fully working example, including all the necessary boilerplate code, can be found in the SnarkyJS repository in the examples folder [here](https://github.com/o1-labs/snarkyjs/tree/main/src/examples/zkapps/merkle_tree).
+
+</Alert>
+
+```ts
+class Leaderboard extends SmartContract {
+  // the root it the root hash of our off-chain Merkle tree
+  @state(Field) root = State<Field>();
+
+  // z is the hashed number we want to guess!
+  @state(Field) z = State<Field>();
+
+  deploy(args: DeployArgs) {
+    super.deploy(args);
+
+    // this is our hash we want to guess! its the hash of the preimage "22", but keep it a secret!
+    this.z.set(
+      Field(
+        "17057234437185175411792943285768571642343179330449434169483610110583519635705"
+      )
+    );
+  }
+
+  @method
+  guessPreimage(guess: Field, account: Account, path: MerkleWitness) {
+    // we fetch z from the chain
+    let z = this.z.get();
+    this.z.assertEquals(z);
+
+    // if our guess preimage hashes to our target, we won a point!
+    Poseidon.hash([guess]).assertEquals(z);
+
+    // we fetch the on-chain commitment
+    let commitment = this.commitment.get();
+    this.commitment.assertEquals(commitment);
+
+    // we check that the account is within the committed Merkle Tree
+    path.calculateRoot(account.hash()).assertEquals(commitment);
+
+    // we update the account and grant one point!
+    let newAccount = account.addPoints(1);
+
+    // we calculate the new Merkle Root, based on the account changes
+    let newCommitment = path.calculateRoot(newAccount.hash());
+
+    this.commitment.set(newCommitment);
+  }
+
+  @method
```

Comment:
Has this code been run through Prettier? Our code elsewhere uses `@method` on the same line as the function name. We should follow what our Prettier config does to the code please